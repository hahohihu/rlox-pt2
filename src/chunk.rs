use num_enum::{FromPrimitive, IntoPrimitive};

use crate::{ui::Span, value::Value};

#[derive(Debug, Eq, PartialEq, FromPrimitive, IntoPrimitive)]
#[repr(u8)]
pub enum OpCode {
    // uncommented opcodes have 0 follow bytes
    Return,
    Constant, // 1 follow byte for a constant index
    Negate,
    Add,
    Sub,
    Mul,
    Div,
    #[num_enum(default)]
    Invalid,
}

#[derive(Default, Debug, Clone)]
pub struct Chunk {
    // INVARIANT: An OpCode must be followed by however many bytes are specified
    pub instructions: Vec<u8>,
    pub spans: Vec<Span>,
    pub constants: Vec<Value>,
}

impl Chunk {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_constant(&mut self, value: Value) -> u8 {
        self.constants.push(value);
        let index = self.constants.len() - 1;
        index.try_into().expect("Too many constants")
    }

    pub fn disassemble(&self, name: &str, source: &str) {
        println!("==== {name} ====");
        let mut i = 0;
        while i < self.instructions.len() {
            i = self.disassemble_instruction(i, source);
        }
    }

    /// SAFETY: OpCode invariants must be upheld. If an opcode is n bytes, n bytes _must_ be inserted
    pub unsafe fn write_byte(&mut self, byte: impl Into<u8>, origin: Span) {
        self.instructions.push(byte.into());
        self.spans.push(origin);
    }

    fn simple_instruction(name: &str, offset: &mut usize) {
        println!("{name}");
        *offset += 1;
    }

    fn constant_instruction(&self, name: &str, offset: &mut usize) {
        let index = self.instructions[*offset + 1];
        let value = self.constants[index as usize];
        println!("{name:<16} {index:>4} '{value}'");
        *offset += 2;
    }

    pub fn disassemble_instruction(&self, mut offset: usize, source: &str) -> usize {
        print!("{:0>4} ", offset);
        if offset > 0 && self.spans[offset] == self.spans[offset - 1] {
            print!("{:<8}", "|");
        } else {
            print!("{:<8}", &source[self.spans[offset]]);
        }

        let chunk = self.instructions[offset];
        let instruction: OpCode = chunk.into();
        match instruction {
            OpCode::Return => Chunk::simple_instruction("RETURN", &mut offset),
            OpCode::Constant => self.constant_instruction("CONSTANT", &mut offset),
            OpCode::Negate => Chunk::simple_instruction("NEGATE", &mut offset),
            OpCode::Add => Chunk::simple_instruction("ADD", &mut offset),
            OpCode::Sub => Chunk::simple_instruction("SUBTRACT", &mut offset),
            OpCode::Mul => Chunk::simple_instruction("MULTIPLY", &mut offset),
            OpCode::Div => Chunk::simple_instruction("DIVIDE", &mut offset),
            OpCode::Invalid => {
                println!("INVALID OPCODE: {chunk}");
                offset += 1;
            }
        }
        offset
    }
}
