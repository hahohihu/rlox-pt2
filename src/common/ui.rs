use std::{fmt::Display, ops::Index};

use arbitrary::Arbitrary;
use tracing::warn;

#[derive(Copy, Clone, PartialEq, Eq, Debug, Arbitrary)]
pub struct Span {
    begin: u32,
    end: u32,
}

impl From<logos::Span> for Span {
    fn from(value: logos::Span) -> Self {
        Self {
            begin: value.start.try_into().unwrap(),
            end: value.end.try_into().unwrap(),
        }
    }
}

impl Index<Span> for str {
    type Output = str;
    fn index(&self, index: Span) -> &Self::Output {
        &self[index.begin as usize..index.end as usize]
    }
}

impl Span {
    pub fn unite(self, other: Span) -> Span {
        Span {
            begin: self.begin.min(other.begin),
            end: self.end.max(other.end),
        }
    }

    pub fn unite_many(spans: &[Span]) -> Span {
        debug_assert!(!spans.is_empty());
        if let Some(span) = spans.get(0) {
            spans[1..].iter().fold(*span, |a, b| a.unite(*b))
        } else {
            warn!("Empty set of spans should never happen");
            Span::from(0..0)
        }
    }

    pub fn maybe_unite<T>(self, other: Option<Spanned<T>>) -> Span {
        if let Some(span) = other {
            self.unite(span.span)
        } else {
            self
        }
    }
}

impl ariadne::Span for Span {
    type SourceId = ();
    fn source(&self) -> &Self::SourceId {
        &()
    }

    fn start(&self) -> usize {
        self.begin as usize
    }

    fn end(&self) -> usize {
        self.end as usize
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Arbitrary)]
pub struct Spanned<T> {
    pub data: T,
    pub span: Span,
}

impl<T> From<Spanned<T>> for Span {
    fn from(value: Spanned<T>) -> Self {
        value.span
    }
}

impl<T: Copy> Copy for Spanned<T> {}

impl<T> Spanned<T> {
    pub fn new(data: T, span: Span) -> Self {
        Self { data, span }
    }

    pub fn boxed(self) -> Box<Self> {
        Box::new(self)
    }
}

impl<T: Display> Display for Spanned<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.data.fmt(f)
    }
}

pub trait WithSpanExt: Sized {
    fn with_span(self, span: Span) -> Spanned<Self>;
}

impl<T> WithSpanExt for T {
    fn with_span(self, span: Span) -> Spanned<Self> {
        Spanned { data: self, span }
    }
}

pub const OFFSET: usize = 12;
