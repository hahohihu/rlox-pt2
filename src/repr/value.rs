use std::fmt::Display;

use super::object::Object;

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Value {
    Num(f64),
    Bool(bool),
    Nil,
    Object(Object),
}

impl Display for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Num(n) => n.fmt(f),
            Self::Bool(b) => b.fmt(f),
            Self::Nil => write!(f, "nil"),
            Self::Object(obj) => obj.fmt(f),
        }
    }
}

impl Value {
    pub fn typename(&self) -> &'static str {
        match self {
            Self::Bool(_) => "boolean",
            Self::Num(_) => "number",
            Self::Nil => "nil",
            Self::Object(obj) => obj.typename(),
        }
    }

    pub fn falsey(&self) -> bool {
        matches!(self, Self::Bool(false) | Self::Nil)
    }
}

impl From<bool> for Value {
    fn from(value: bool) -> Self {
        Self::Bool(value)
    }
}

impl From<f64> for Value {
    fn from(value: f64) -> Self {
        Self::Num(value)
    }
}

impl From<&str> for Value {
    fn from(value: &str) -> Self {
        Self::Object(Object::make_str(String::from(value)))
    }
}
