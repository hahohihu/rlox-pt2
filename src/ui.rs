pub type Span = logos::Span;
#[derive(Clone, Debug)]
pub struct Spanned<T> {
    pub data: T,
    pub span: Span,
}

impl<T> Spanned<T> {
    pub fn new(data: T, span: Span) -> Self {
        Self { data, span }
    }
}