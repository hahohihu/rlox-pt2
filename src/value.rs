use std::fmt::Display;

#[derive(Clone, Copy, PartialEq, PartialOrd, Debug)]
pub enum Value {
    Num(f64),
    Bool(bool),
    Nil,
}

impl Display for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Num(n) => n.fmt(f),
            Self::Bool(b) => b.fmt(f),
            Self::Nil => write!(f, "nil"),
        }
    }
}

impl Value {
    pub fn typename(&self) -> &'static str {
        match self {
            Self::Bool(_) => "boolean",
            Self::Num(_) => "number",
            Self::Nil => "nil",
        }
    }
}