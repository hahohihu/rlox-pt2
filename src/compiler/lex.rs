use crate::common::ui::{Span, Spanned};
use logos::Logos;

#[derive(Logos, Debug, PartialEq, Copy, Clone)]
#[logos(skip r"[ \n\r\f]+")]
pub enum Token {
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token(",")]
    Comma,
    #[token(".")]
    Dot,
    #[token("-")]
    Minus,
    #[token("+")]
    Plus,
    #[token(";")]
    Semicolon,
    #[token("/")]
    Slash,
    #[token("*")]
    Star,

    #[token("!")]
    Bang,
    #[token("!=")]
    BangEq,
    #[token("=")]
    Eq,
    #[token("==")]
    EqEq,
    #[token(">")]
    Greater,
    #[token(">=")]
    GreaterEq,
    #[token("<")]
    Less,
    #[token("<=")]
    LessEq,

    #[regex("[a-zA-Z_][a-zA-Z_0-9]*")]
    Ident,
    #[regex(r#""[^"]*""#)]
    String,
    // leading digit is mandatory, possibly incongruent with book
    #[regex("[0-9]+(\\.[0-9]+)?")]
    Num,
    #[token("true")]
    True,
    #[token("false")]
    False,

    #[token("and")]
    And,
    #[token("class")]
    Class,
    #[token("else")]
    Else,
    #[token("for")]
    For,
    #[token("fun")]
    Fun,
    #[token("if")]
    If,
    #[token("nil")]
    Nil,
    #[token("or")]
    Or,
    #[token("print")]
    Print,
    #[token("return")]
    Return,
    #[token("super")]
    Super,
    #[token("this")]
    This,
    #[token("var")]
    Var,
    #[token("while")]
    While,
    Eof,
}

pub struct Lexer<'src>(logos::SpannedIter<'src, Token>);

impl<'src> Lexer<'src> {
    pub fn new(src: &'src str) -> Self {
        Self(Token::lexer(src).spanned())
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Result<Spanned<Token>, Span>;

    fn next(&mut self) -> Option<Self::Item> {
        self.0.next().map(|(tok, span)| {
            let span = span.into();
            tok.map(|t| Spanned::new(t, span)) // rustfmt guard
                .map_err(|_| span)
        })
    }
}

#[cfg(test)]
mod tests {
    use super::{Lexer, Token};
    use crate::common::ui::Span;
    use Token::*;

    fn lex(src: &str) -> Result<Vec<Token>, Span> {
        Lexer::new(src)
            .map(|t| t.map(|t| t.data))
            .collect::<Result<Vec<_>, _>>()
    }

    fn lex_ok(src: &str) -> Vec<Token> {
        match lex(src) {
            Ok(t) => t,
            Err(s) => panic!("Mismatch at {:?} on {}", s, &src[s]),
        }
    }

    #[test]
    fn bools() {
        assert_eq!(lex_ok("true false"), &[True, False]);
    }

    #[test]
    fn comparisons() {
        assert_eq!(lex_ok("< <= == ="), &[Less, LessEq, EqEq, Eq]);
    }

    #[test]
    fn multiline_strings() {
        assert_eq!(lex_ok("\"1\n2\n3\n\""), &[String])
    }

    #[test]
    fn sequence_of_numbers() {
        assert_eq!(lex_ok("1 1"), &[Num, Num]);
    }
}
