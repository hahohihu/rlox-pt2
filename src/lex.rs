use crate::ui::{Span, Spanned};
use logos::Logos;

#[derive(Logos, Debug, PartialEq, Copy, Clone)]
#[logos(skip r"[ \n\r\f]+")]
pub enum Token<'src> {
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token(",")]
    Comma,
    #[token(".")]
    Dot,
    #[token("-")]
    Minus,
    #[token("+")]
    Plus,
    #[token(";")]
    Semicolon,
    #[token("/")]
    Slash,
    #[token("*")]
    Star,

    #[token("!")]
    Bang,
    #[token("!=")]
    BangEq,
    #[token("=")]
    Eq,
    #[token("==")]
    EqEq,
    #[token(">")]
    Greater,
    #[token(">=")]
    GreaterEq,
    #[token("<")]
    Less,
    #[token("<=")]
    LessEq,

    #[regex("[a-zA-Z_][a-zA-Z_0-9]*")]
    Ident(&'src str),
    #[regex(r#""[^"]*""#, |lex| {
        let string = lex.slice();
        &string[1..string.len() - 1] // trim "
    })]
    String(&'src str),
    // leading digit is mandatory, possibly incongruent with book
    #[regex("[0-9]+(.[0-9]+)", |lex| lex.slice().parse().ok())]
    Num(f64),
    #[regex("true|false", |lex| lex.slice() == "true")]
    Bool(bool),

    #[token("and")]
    And,
    #[token("class")]
    Class,
    #[token("else")]
    Else,
    #[token("for")]
    For,
    #[token("fun")]
    Fun,
    #[token("if")]
    If,
    #[token("nil")]
    Nil,
    #[token("or")]
    Or,
    #[token("print")]
    Print,
    #[token("return")]
    Return,
    #[token("super")]
    Super,
    #[token("this")]
    This,
    #[token("var")]
    Var,
    #[token("while")]
    While,
}

pub struct Lexer<'src>(logos::SpannedIter<'src, Token<'src>>);

impl<'src> Lexer<'src> {
    pub fn new(src: &'src str) -> Self {
        Self(Token::lexer(src).spanned())
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Result<Spanned<Token<'src>>, Span>;

    fn next(&mut self) -> Option<Self::Item> {
        self.0.next().map(|(tok, span)| {
            tok.map(|t| Spanned::new(t, span)) // rustfmt guard
                .map_err(|_| span.clone())
        })
    }
}

#[cfg(test)]
mod tests {
    use super::{Lexer, Token};
    use Token::*;

    fn lex(src: &str) -> Result<Vec<Token>, logos::Span> {
        Lexer::new(src)
            .map(|t| t.map(|t| t.data))
            .collect::<Result<Vec<_>, _>>()
    }

    fn lex_ok(src: &str) -> Vec<Token> {
        match lex(src) {
            Ok(t) => t,
            Err(s) => panic!("Mismatch at {:?} on {}", s.clone(), &src[s]),
        }
    }

    #[test]
    fn bools() {
        assert_eq!(lex_ok("true false"), &[Bool(true), Bool(false)]);
    }

    #[test]
    fn comparisons() {
        assert_eq!(lex_ok("< <= == ="), &[Less, LessEq, EqEq, Eq]);
    }

    #[test]
    fn multiline_strings() {
        assert_eq!(lex_ok("\"1\n2\n3\n\""), &[String("1\n2\n3\n")])
    }
}
