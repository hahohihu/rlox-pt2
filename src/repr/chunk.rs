use std::io::Write;

use num_enum::{FromPrimitive, IntoPrimitive};

use super::{interner::Interner, value::Value};
use crate::common::ui::Span;

#[derive(Debug, Eq, PartialEq, FromPrimitive, IntoPrimitive)]
#[repr(u8)]
pub enum OpCode {
    // 0 follow bytes ====
    Return,
    Nil,
    True,
    False,
    // 1 follow bytes ====
    Constant, // 1: a constant index
    // No follow bytes but data-dependent
    // Unary
    Negate,
    Not,
    Print,
    Pop,
    GetGlobal,
    DefineGlobal,
    SetGlobal,
    // Binary
    Add,
    Sub,
    Mul,
    Div,
    Equal,
    Greater,
    Less,
    #[num_enum(default)]
    Invalid,
}

#[derive(Default, Debug, Clone)]
pub struct Chunk {
    // INVARIANT: An OpCode must be followed by however many bytes are specified
    pub instructions: Vec<u8>,
    pub spans: Vec<Span>,
    // Owned by this
    constants: Vec<Value>,
    // Owned by this
    pub globals: Interner,
}

impl Drop for Chunk {
    fn drop(&mut self) {
        for constant in &self.constants {
            if let Value::Object(obj) = constant {
                unsafe {
                    // SAFETY: See safety invariant on constants
                    obj.free();
                }
            }
        }
    }
}

impl Chunk {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_constant(&mut self, value: Value) -> u8 {
        self.constants.push(value);
        let index = self.constants.len() - 1;
        index.try_into().expect("Too many constants")
    }

    pub fn get_constant(&self, index: u8) -> Value {
        self.constants[index as usize]
    }

    /// SAFETY: OpCode invariants must be upheld. If an opcode is n bytes, n bytes _must_ be inserted
    pub unsafe fn write_byte(&mut self, byte: impl Into<u8>, origin: Span) {
        self.instructions.push(byte.into());
        self.spans.push(origin);
    }

    pub fn disassemble(&self, name: &str, source: &str, mut stdout: impl Write) {
        writeln!(stdout, "==== {name} ====").unwrap();
        let mut i = 0;
        while i < self.instructions.len() {
            i = self.disassemble_instruction(i, source, &mut stdout);
        }
    }

    fn simple_instruction(name: &str, offset: &mut usize, mut stdout: impl Write) {
        writeln!(stdout, "{name}").unwrap();
        *offset += 1;
    }

    fn constant_instruction(&self, name: &str, offset: &mut usize, mut stdout: impl Write) {
        let index = self.instructions[*offset + 1];
        let value = self.constants[index as usize];
        writeln!(stdout, "{name:<16} {index:>4} '{value}'").unwrap();
        *offset += 2;
    }

    fn global_instruction(&self, name: &str, offset: &mut usize, mut stdout: impl Write) {
        let index = self.instructions[*offset + 1];
        let value = self.globals.get_name(index);
        writeln!(stdout, "{name:<16} {index:>4} '{value}'").unwrap();
        *offset += 2;
    }

    pub fn disassemble_instruction(
        &self,
        mut offset: usize,
        source: &str,
        mut stdout: impl Write,
    ) -> usize {
        write!(stdout, "{:0>4} ", offset).unwrap();
        if offset > 0 && self.spans[offset] == self.spans[offset - 1] {
            write!(stdout, "{:<8}", "|").unwrap();
        } else {
            write!(stdout, "{:<8}", &source[self.spans[offset]]).unwrap();
        }

        let chunk = self.instructions[offset];
        let instruction: OpCode = chunk.into();
        let mut simple = |str| Chunk::simple_instruction(str, &mut offset, &mut stdout);
        match instruction {
            OpCode::Return => simple("RETURN"),
            OpCode::Constant => self.constant_instruction("CONSTANT", &mut offset, stdout),
            OpCode::Negate => simple("NEGATE"),
            OpCode::Add => simple("ADD"),
            OpCode::Sub => simple("SUBTRACT"),
            OpCode::Mul => simple("MULTIPLY"),
            OpCode::Div => simple("DIVIDE"),
            OpCode::Nil => simple("NIL"),
            OpCode::Not => simple("NOT"),
            OpCode::True => simple("TRUE"),
            OpCode::False => simple("FALSE"),
            OpCode::Equal => simple("EQUAL"),
            OpCode::Greater => simple("GREATER"),
            OpCode::Less => simple("LESS"),
            OpCode::Print => simple("PRINT"),
            OpCode::Pop => simple("POP"),
            OpCode::DefineGlobal => self.global_instruction("DEFINE_GLOBAL", &mut offset, stdout),
            OpCode::GetGlobal => self.global_instruction("GET_GLOBAL", &mut offset, stdout),
            OpCode::SetGlobal => self.global_instruction("SET_GLOBAL", &mut offset, stdout),
            OpCode::Invalid => {
                writeln!(stdout, "INVALID OPCODE: {chunk}").unwrap();
                offset += 1;
            }
        }
        offset
    }
}
