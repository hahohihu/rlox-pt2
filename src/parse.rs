use std::iter::Peekable;

#[cfg(feature = "verbose_parsing")]
use tracing::trace;
#[cfg(not(feature = "verbose_parsing"))]
use crate::noop as trace;

use crate::chunk::Chunk;
use crate::chunk::OpCode;
use crate::lex::Lexer;
use crate::lex::Token;

use crate::ui;
use crate::ui::*;
use crate::value::Value;

struct Parser<'src> {
    lexer: Peekable<Lexer<'src>>,
    source: &'src str,
}

macro_rules! expect {
    ($self:expr, $msg:expr, $pat:pat => $ret:expr) => {{
        let tok = $self.pop(stringify!($pat))?;
        match tok.data {
            $pat => Ok(Spanned::new($ret, tok.span)),
            _ => return Err(ParseError::ExpectError {
                expected: $msg,
                got: tok.span
            }),
        }
    }};
    ($self:expr, $msg:expr, $pat:pat) => {
        expect!($self, $msg, $pat => ())
    }
}

/// The same rules around emit_byte applies
macro_rules! emit_bytes {
    ($chunk:expr, $span:expr; $($byte:expr),+ $(,)?) => {{
        let span: Span = $span.into();
        $($chunk.emit_byte($byte, span);)+
    }}
}

impl Chunk {
    /// PRECONDITION: If an OpCode is emitted, it must have the specified number of follow bytes + follow other constraints
    unsafe fn emit_byte(&mut self, byte: impl Into<u8>, span: impl Into<Span>) {
        self.write_byte(byte, span.into());
    }

    fn emit_constant(&mut self, value: Value, span: impl Into<Span>) {
        let constant = self.add_constant(value);
        // SAFETY: constant
        unsafe { emit_bytes!(self, span; OpCode::Constant, constant) }
    }

    fn emit_return(&mut self) {
        unsafe {
            // SAFETY: constant
            self.emit_byte(OpCode::Return, 0..0);
        }
    }
}

#[derive(Debug)]
pub enum ParseError {
    EOF { expected: &'static str },
    InvalidToken(Span),
    ExpectError { expected: &'static str, got: Span },
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
enum Precedence {
    None,
    Start,
    Assignment,
    Or,
    And,
    Equality,
    Comparison,
    Term,
    Factor,
    Unary,
    Call,
    Primary,
}

impl From<Token> for Precedence {
    fn from(value: Token) -> Self {
        match value {
            Token::Minus => Self::Term,
            Token::Plus => Self::Term,
            Token::Slash => Self::Factor,
            Token::Star => Self::Factor,
            Token::EqEq => Self::Equality,
            Token::Greater => Self::Comparison,
            Token::GreaterEq => Self::Comparison,
            Token::Less => Self::Comparison,
            Token::LessEq => Self::Comparison,
            _ => Self::None,
        }
    }
}

impl ParseError {
    pub fn print(&self, source: &str) {
        use ariadne::{Color, Label, Report, ReportKind, Source};
        let report = match self {
            Self::InvalidToken(span) => Report::build(ReportKind::Error, (), ui::OFFSET)
                .with_message("Lexing error")
                .with_label(
                    Label::new(*span)
                        .with_color(Color::Red)
                        .with_message("Invalid token"),
                ),
            Self::EOF { expected } => {
                let end = source.len().saturating_sub(1);
                Report::build(ReportKind::Error, (), ui::OFFSET)
                    .with_message("Unexpected end of file")
                    .with_label(
                        Label::new(Span::from(end..source.len()))
                            .with_color(Color::Green)
                            .with_message(format!("Expected {expected}")),
                    )
            }
            Self::ExpectError { expected, got } => Report::build(ReportKind::Error, (), ui::OFFSET)
                .with_message("Parse error")
                .with_label(
                    Label::new(*got)
                        .with_color(Color::Green)
                        .with_message(format!("Expected {expected}")),
                ),
        };
        report.finish().eprint(Source::from(source)).unwrap();
    }
}

pub type ParseRes<T> = Result<T, ParseError>;

impl<'src> Parser<'src> {
    fn new(source: &'src str) -> Self {
        let lexer = Lexer::new(source).peekable();
        Self { lexer, source }
    }

    fn pop(&mut self, expected: &'static str) -> ParseRes<Spanned<Token>> {
        match self.lexer.next() {
            Some(Ok(t)) => {
                trace!("popping '{}'", &self.source[t.span]);
                Ok(t)
            }
            Some(Err(t)) => Err(ParseError::InvalidToken(t)),
            None => Err(ParseError::EOF { expected }),
        }
    }

    fn peek(&mut self) -> Option<ParseRes<Spanned<Token>>> {
        let tok = self.lexer.peek()?;
        Some(match tok {
            Ok(t) => Ok(*t),
            Err(t) => Err(ParseError::InvalidToken(*t)),
        })
    }

    /// ensures: Ok(_) ==> Exactly one additional value on the stack
    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, chunk)))]
    fn primary(&mut self, chunk: &mut Chunk) -> Result<(), ParseError> {
        let token = self.pop("primary")?;
        match token.data {
            Token::Minus => {
                self.primary(chunk)?;
                // SAFETY: a value must be on the stack for negate to work, which primary guarantees
                unsafe { chunk.emit_byte(OpCode::Negate, token.span) }
            }
            Token::Bang => {
                self.primary(chunk)?;
                // SAFETY: a value must be on the stack for negate to work, which primary guarantees
                unsafe { chunk.emit_byte(OpCode::Not, token.span) }
            }
            Token::LParen => {
                self.expression(chunk, Precedence::Start)?;
                expect!(self, ") after expression", Token::RParen)?;
            }
            Token::Num => {
                let n = self.source[token.span].parse::<f64>().unwrap();
                chunk.emit_constant(Value::Num(n), token.span)
            }
            Token::True => unsafe {
                // SAFETY: constant
                chunk.emit_byte(OpCode::True, token.span);
            },
            Token::False => unsafe {
                // SAFETY: constant
                chunk.emit_byte(OpCode::False, token.span);
            },
            Token::Nil => unsafe {
                // SAFETY: constant
                chunk.emit_byte(OpCode::Nil, token.span);
            },
            Token::String => {
                // remove parens
                let str = &self.source[token.span][1..];
                let str = &str[..str.len() - 1];
                chunk.emit_constant(
                    Value::from(str),
                    token.span,
                )
            }
            _ => todo!(),
        }
        Ok(())
    }

    /// ensures: Ok(_) ==> Exactly one additional value on the stack
    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, chunk)))]
    fn expression(&mut self, chunk: &mut Chunk, min: Precedence) -> Result<(), ParseError> {
        self.primary(chunk)?;
        loop {
            let Some(token) = self.peek() else {
                break;
            };
            let operation = token?;

            let prec = Precedence::from(operation.data); // todo: everything left associative
            if prec < min {
                break;
            }

            self.pop("unreachable").unwrap();
            self.expression(chunk, prec)?;

            // SAFETY: There must be 2 prior values on the stack. This is guaranteed by primary + expression
            unsafe {
                macro_rules! emit {
                    ($($opcode:expr),+) => {
                        emit_bytes!(chunk, operation.span; $($opcode,)+)
                    };
                }
                match operation.data {
                    Token::Minus => emit!(OpCode::Sub),
                    Token::Plus => emit!(OpCode::Add),
                    Token::Slash => emit!(OpCode::Div),
                    Token::Star => emit!(OpCode::Mul),
                    Token::EqEq => emit!(OpCode::Equal),
                    Token::BangEq => emit!(OpCode::Equal, OpCode::Not),
                    Token::Greater => emit!(OpCode::Greater),
                    Token::GreaterEq => emit!(OpCode::Less, OpCode::Not),
                    Token::Less => emit!(OpCode::Less),
                    Token::LessEq => emit!(OpCode::Greater, OpCode::Not),
                    _ => unreachable!("Precedence::from must handle this"),
                }
            }
        }
        Ok(())
    }

    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, chunk)))]
    fn statement(&mut self, chunk: &mut Chunk) -> Result<(), ParseError> {
        let tok = self.pop("statement")?;
        let opcode = match tok.data {
            Token::Print => OpCode::Print,
            Token::Return => OpCode::Return,
            _ => return Err(ParseError::ExpectError {
                expected: "statement",
                got: tok.span
            })
        };
        self.expression(chunk, Precedence::Start)?;
        unsafe {
            chunk.emit_byte(opcode, tok.span);
        }
        Ok(())
    }
}

pub fn compile(source: &str) -> Result<Chunk, ParseError> {
    let mut chunk = Chunk::new();
    let mut parser = Parser::new(source);
    parser.statement(&mut chunk)?;
    if let Some(t) = parser.peek() {
        let t = t?;
        return Err(ParseError::ExpectError {
            expected: "EOF",
            got: t.span,
        });
    }
    chunk.emit_return();
    Ok(chunk)
}
