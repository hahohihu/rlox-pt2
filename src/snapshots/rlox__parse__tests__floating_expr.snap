---
source: src/parse.rs
expression: "crate::util::mock_interpret(\"1;\")"
---
stdout:


stderr:
Error: Parse error
   ╭─[<unknown>:?:?]
   │
 1 │ 1;
   │ ┬  
   │ ╰── Expected statement
───╯


